DeviceAdministration/Infrastructure/Repository/DeviceRegistryRepository.cs

```diff
@@ -108,29 +49,31 @@ public DeviceRegistryRepository(IConfigurationProvider configProvider, IDocDbRes
         /// <returns></returns>
         public async Task<DeviceModel> AddDeviceAsync(DeviceModel device)
         {
-            string deviceId = device.DeviceProperties.DeviceID;
-            DeviceModel existingDevice = await this.GetDeviceAsync(deviceId);
+            if (string.IsNullOrEmpty(device.id))
+            {
+                device.id = Guid.NewGuid().ToString();
+            }
 
+            var id = device.id;
+            DeviceModel existingDevice = await GetDeviceAsync(device.DeviceProperties.DeviceID);
             if (existingDevice != null)
             {
-                throw new DeviceAlreadyRegisteredException(deviceId);
+                throw new DeviceAlreadyRegisteredException(device.DeviceProperties.DeviceID);
             }
 
-            device = (await _docDbRestUtil.SaveNewDocumentAsync<DeviceModel>(device)).ToObject<DeviceModel>();
-
+            await _documentClient.SaveAsync(id, device);
```

Comment:
using id as the key is something new we're introducing here? Wouldn't this break the existing users ?